File: wrappers/scala/src/main/scala/com/turbotoken/TurboToken.scala
package com.turbotoken

import java.util.concurrent.ConcurrentHashMap

/** Main entry point for the turbotoken Scala API.
  *
  * {{{
  * val enc = TurboToken.getEncoding("cl100k_base")
  * val tokens = enc.encode("hello world")
  * val decoded = enc.decode(tokens)
  *
  * // Or by model name:
  * val enc2 = TurboToken.getEncodingForModel("gpt-4o")
  * }}}
  */
object TurboToken {

  private val cache = new ConcurrentHashMap[String, Encoding]()

  /** Returns the native library version string. */
  def version: String = com.turbotoken.TurboToken.version()

  /** Returns an Encoding for the given encoding name (e.g. "cl100k_base", "o200k_base").
    * Encoding instances are cached and reused.
    *
    * @throws UnknownEncodingException if the encoding name is unknown
    */
  def getEncoding(name: String): Encoding = {
    cache.computeIfAbsent(name, { k =>
      // Validate the name exists in our registry
      Registry.getEncodingSpec(k)
      // Delegate to Java for actual native loading
      val javaEnc = com.turbotoken.TurboToken.getEncoding(k)
      new Encoding(javaEnc)
    })
  }

  /** Returns an Encoding for the given model name (e.g. "gpt-4o", "gpt-3.5-turbo").
    *
    * @throws UnknownModelException if the model cannot be mapped to an encoding
    */
  def getEncodingForModel(model: String): Encoding = {
    val encodingName = Registry.modelToEncoding(model)
    getEncoding(encodingName)
  }

  /** Returns a sorted list of all supported encoding names. */
  def listEncodingNames: Seq[String] = Registry.listEncodingNames
}

File: wrappers/kotlin/src/main/kotlin/com/turbotoken/TurboToken.kt
@file:JvmName("KTurboToken")
package com.turbotoken

import java.util.concurrent.ConcurrentHashMap

/**
 * Main entry point for the turbotoken Kotlin API.
 *
 * ```kotlin
 * val enc = TurboTokenK.getEncoding("cl100k_base")
 * val tokens = enc.encode("hello world")
 * val decoded = enc.decode(tokens)
 *
 * // Or by model name:
 * val enc2 = TurboTokenK.getEncodingForModel("gpt-4o")
 *
 * // Extension functions:
 * val count = "hello world".tokenCount(enc)
 * ```
 */
object TurboTokenK {

    private val cache = ConcurrentHashMap<String, KEncoding>()

    /**
     * Returns the native library version string.
     */
    fun version(): String = KNativeBridge.version()

    /**
     * Returns a KEncoding for the given encoding name.
     * Encoding instances are cached and reused.
     * @throws IllegalArgumentException if the encoding name is unknown
     */
    fun getEncoding(name: String): KEncoding =
        cache.computeIfAbsent(name) {
            KEncoding(TurboToken.getEncoding(it))
        }

    /**
     * Returns a KEncoding for the given model name.
     * @throws IllegalArgumentException if the model cannot be mapped to an encoding
     */
    fun getEncodingForModel(model: String): KEncoding {
        val encodingName = EncodingRegistry.modelToEncoding(model)
        return getEncoding(encodingName)
    }

    /**
     * Returns a sorted list of all supported encoding names.
     */
    fun listEncodingNames(): List<String> = EncodingRegistry.listEncodingNames()

    companion object {
        /**
         * Static access for Java interop.
         */
        @JvmStatic
        fun getEncodingStatic(name: String): KEncoding = getEncoding(name)

        @JvmStatic
        fun getEncodingForModelStatic(model: String): KEncoding = getEncodingForModel(model)

        @JvmStatic
        fun listEncodingNamesStatic(): List<String> = listEncodingNames()

        @JvmStatic
        fun versionStatic(): String = version()
    }
}
